R/zchunk_LA1321.Cement_china.R

```diff
@@ -0,0 +1,117 @@
+#' @param ... other optional parameters, depending on command
+#' @return Depends on \code{command}: either a vector of required inputs,
+#' a vector of output names, or (if \code{command} is "MAKE") all
+#' the generated outputs: \code{L1321.out_Mt_province_cement_Yh}, \code{L1321.IO_GJkg_province_cement_F_Yh}, \code{L1321.in_EJ_province_cement_F_Y}. The corresponding file in the
+#' original data system was \code{LA1321.Cement.R} (gcam-china level1).
+#' @details The tables for cement production, i.e., out, and energy inputs, i.e., in, were calculated by applying state shares to national data.
+#' @details The state shares were determined by the states' relative values of cement shipments.
+#' @details The input-out coefficients were downscaled to the states in proportation to the national data.
+#' @importFrom assertthat assert_that
+#' @importFrom dplyr filter mutate select
+#' @importFrom tidyr gather spread
+#' @author Yang Aug 2018
+module_gcam.china_LA1321.Cement <- function(command, ...) {
+  if(command == driver.DECLARE_INPUTS) {
+    return(c("L101.inNBS_Mtce_province_S_F",
+             "L1321.out_Mt_R_cement_Yh",
+             "L1321.IO_GJkg_R_cement_F_Yh",
+             "L1321.in_EJ_R_cement_F_Y"))
+  } else if(command == driver.DECLARE_OUTPUTS) {
+    return(c("L1321.out_Mt_province_cement_Yh",
+             "L1321.IO_GJkg_province_cement_F_Yh",
+             "L1321.in_EJ_province_cement_F_Y"))
+  } else if(command == driver.MAKE) {
+
+    all_data <- list(...)[[1]]
+
+    sector <- fuel <- province <- multiplier <- GCAM_region_ID <- year <- value <- NULL   # silence package check notes
+
+    # Load required inputs
+    L1321.out_Mt_R_cement_Yh <- get_data(all_data, "L1321.out_Mt_R_cement_Yh")
+    L1321.IO_GJkg_R_cement_F_Yh <- get_data(all_data, "L1321.IO_GJkg_R_cement_F_Yh")
+    L1321.in_EJ_R_cement_F_Y <- get_data(all_data, "L1321.in_EJ_R_cement_F_Y")
+
+    # Proxy for downscaling cement to provinces
+    # TODO: Find a better proxy
+    L101.inNBS_Mtce_province_S_F <- get_data(all_data, "L101.inNBS_Mtce_province_S_F")
+
+    # ===================================================
+    # Assigning national cement production to provinces
+    L101.inNBS_Mtce_province_S_F %>%
+      filter(sector == "industry" & fuel == "coal") %>%
+      group_by(year) %>%
+      mutate(multiplier = value /sum(value)) %>%
+      ungroup %>%
+      group_by(province) %>%
+      mutate(multiplier = approx_fun(year, multiplier, rule = 2), sector = "cement") %>%
+      ungroup() %>%
+      select(-value, -fuel) ->
+      L1321.in_pct_province_cement
+
+    # This section is calculating province-level data by multiplying the province share by the china component in the global data
+    # This will generate an output table: Cement production by province / historical year
+    L1321.in_pct_province_cement %>%
+      left_join_error_no_match(L1321.out_Mt_R_cement_Yh %>% filter(GCAM_region_ID == 11), by = c("year", "sector")) %>%
```

Comment:
Here and throughout, let's use a constant `gcam.CHINA_CODE` defined in `constants.R` please, not a numeric code. (See lines 35-36 in that file.)